cni/pkg/nodeagent/ztunnelserver.go

```diff
@@ -0,0 +1,480 @@
+		// Anything other than "file not found" is an error.
+		return nil, fmt.Errorf("failed to remove unix://%s: %w", addr, err)
+	}
+
+	l, err := net.ListenUnix("unixpacket", resolvedAddr)
+	if err != nil {
+		return nil, fmt.Errorf("failed to listen unix: %w", err)
+	}
+
+	return &ztunnelServer{
+		listener: l,
+		conns: &connMgr{
+			connectionSet: map[*ZtunnelConnection]struct{}{},
+		},
+		pods: pods,
+	}, nil
+}
+
+func (z *ztunnelServer) Close() error {
+	return z.listener.Close()
+}
+
+// this is used in tests
+// nolint: unused
+func (z *ztunnelServer) numConns() int {
+	return z.conns.len()
+}
+
+func (z *ztunnelServer) Run(ctx context.Context) {
+	go func() {
+		<-ctx.Done()
+		z.Close()
+	}()
+
+	for {
+		log.Debug("accepting conn")
+		conn, err := z.accept()
+		if err != nil {
+			if errors.Is(err, net.ErrClosed) {
+				log.Debug("listener closed - returning")
+				return
+			}
+
+			log.Errorf("failed to accept conn: %v", err)
+			continue
+		}
+		log.Debug("connection accepted")
+		go func() {
+			log.Debug("handling conn")
+			if err := z.handleConn(ctx, conn); err != nil {
+				log.Errorf("failed to handle conn: %v", err)
+			}
+		}()
+	}
+}
+
+// ZDS protocol is very simple, for every message sent, and ack is sent.
+// the ack only has temporal correlation (i.e. it is the first and only ack msg after the message was sent)
+// All this to say, that we want to make sure that message to ztunnel are sent from a single goroutine
+// so we don't mix messages and acks.
+// nolint: unparam
+func (z *ztunnelServer) handleConn(ctx context.Context, conn *ZtunnelConnection) error {
+	defer conn.Close()
+	go func() {
+		<-ctx.Done()
+		log.Debug("context cancelled - closing conn")
+		conn.Close()
+	}()
+
+	// before doing anything, add the connection to the list of active connections
+	z.conns.addConn(conn)
+	defer z.conns.deleteConn(conn)
+
+	log.Debug("sending snapshot to ztunnel")
+	if err := z.sendSnapshot(ctx, conn); err != nil {
+		return err
+	}
+	for {
+		// listen for updates:
+		select {
+		case update, ok := <-conn.Updates:
+			if !ok {
+				log.Debug("update channel closed - returning")
+				return nil
+			}
+			log.Debugf("got update to send to ztunnel")
+			resp, err := conn.sendDataAndWaitForAck(update.Update, update.Fd)
+			if err != nil {
+				log.Errorf("ztunnel acked error: err %v ackErr %s", err, resp.GetAck().GetError())
+			}
+			log.Debugf("ztunnel acked")
+			// Safety: Resp is buffered, so this will not block
+			update.Resp <- updateResponse{
+				err:  err,
+				resp: resp,
+			}
+
+		case <-time.After(ztunnelKeepAliveCheckInterval):
+			// do a short read, just to see if the connection to ztunnel is
+			// still alive. As ztunnel shouldn't send anything unless we send
+			// something first, we expect to get an os.ErrDeadlineExceeded error
+			// here if the connection is still alive.
+			// note that unlike tcp connections, reading is a good enough test here.
+			_, err := conn.readMessage(time.Second / 100)
```

Comment:
@hzxuzhonghu - Because this connection is used for passing FDs, we cannot use a framework like gRPC.
@howardjohn -  This is a unix socket, so we can't have a half-open connection like in tcp sockets that make these kind of tests unreliable.

I'm not aware of prior art for detecting an idle connection as closed without using read or write in native go. We can try and introduce something like `epoll`ing for `EPOLLHUP` in a different goroutine, but IMHO I'm don't think it's worth the extra complexity for something that's need mainly for clean-up.

A second approach would be to move reading to its own go-routing, and continuously read messages until the connection is closed, sending them to a channel. That channel with also be used in  `sendDataAndWaitForAck`.

Personally, given that this is just to clean up, I think either approach is fine. If you prefer the second approach we can use it instead.

Reply:
+1 to what yuval said - this will always work. It's crude, but it isn't incorrect.